apps/emqx_authn/src/emqx_authn_initializer.erl

```diff
@@ -0,0 +1,133 @@
+
+%%--------------------------------------------------------------------
+%% External interface
+%%--------------------------------------------------------------------
+
+start_link() ->
+        start_link(
+          #{name => ?AUTHN,
+            interval => ?DEFAULT_INTERVAL,
+            init_fun => fun(_) -> initialize() end}).
+
+start_link(#{} = Opts) ->
+        gen_statem:start_link(?MODULE, Opts, []).
+
+initialize() ->
+    _ = ?AUTHN:register_providers(emqx_authn:providers()),
+
+    lists:foreach(
+      fun({ChainName, RawAuthConfigs}) ->
+              AuthConfig = emqx_authn:check_configs(RawAuthConfigs),
+              ?AUTHN:initialize_authentication(
+                 ChainName,
+                 AuthConfig)
+      end,
+      chain_configs()).
+
+deinitialize() ->
+    ok = ?AUTHN:deregister_providers(provider_types()).
+
+%%--------------------------------------------------------------------
+%% gen_statem callbacks
+%%--------------------------------------------------------------------
+
+callback_mode() -> state_functions.
+
+init(#{init_fun := Fun,
+       name := Name,
+       interval := Interval} = Data)
+  when
+      is_function(Fun, 1)
+      and is_atom(Name)
+      and is_integer(Interval) ->
+    {ok, waiting_for_process, Data, [{state_timeout, 0, check_process}]}.
+
+waiting_for_process(state_timeout, check_process, #{name := Name,
+                                                   interval := Interval} = Data) ->
+    case whereis(Name) of
+        undefined ->
+            ?tp(info, emqx_authn_initializer_no_process, #{name => Name}),
+            {keep_state_and_data, [{state_timeout, Interval, check_process}]};
+        Pid ->
+            ?tp(info, emqx_authn_initializer_process_registered, #{name => Name}),
+            {next_state, initializing, Data, [{next_event, internal, {initialize, Pid}}]}
+    end.
+
+initializing(internal, {initialize, Pid}, #{init_fun := InitFun, name := Name} = Data) ->
+    _ = InitFun(Pid),
+    Ref = monitor(process, Pid),
+    ?tp(info, emqx_authn_initializer_process_initialized, #{name => Name}),
+    {next_state, idle, Data#{ref => Ref}, [{hibernate, true}]}.
+
+idle(info, {'DOWN', Ref, process, _Pid, _} = Msg, #{ref := Ref, name:= Name} = Data) ->
+    true = demonitor(Ref),
+    NewData = maps:without([ref], Data),
+    ?tp(info, emqx_authn_initializer_process_died, #{name => Name, message => Msg}),
+    {next_state, waiting_for_process, NewData, [{state_timeout, 0, check_process}]}.
+
+%%--------------------------------------------------------------------
+%% Internal functions
+%%--------------------------------------------------------------------
+
+chain_configs() ->
+    [global_chain_config() | listener_chain_configs()].
+
+global_chain_config() ->
+    {?GLOBAL, emqx:get_raw_config([<<"authentication">>], [])}.
+
+listener_chain_configs() ->
+    lists:map(
+     fun({ListenerID, _}) ->
+        {ListenerID, emqx:get_raw_config(auth_config_path(ListenerID), [])}
```

Comment:
Seems that previously only global authentication had its config validated.
Listener configs used configs from `emqx_listeners:list/0` which return already translated config, and there is no authentication schema validation during translation.